### DISCUSSION TITLE
How to MBW for XYZ plot?

### DISCUSSION THREAD

**Discussion by miasik** [wkpark/sd-webui-model-mixer#98] [Q&A] [answered]
I see the option to set MBW in XYZ plot.
What is the format for this field? What is the way to put several MBW there?

**Comment by wkpark** [comment]
> I see the option to set MBW in XYZ plot. What is the format for this field? What is the way to put several MBW there?

when you select MBW alpha B,
you can scan, "MBW alpha" values, like as
"0,0,0,0,0,1,1,1,0,0,0,0.....","0.0.0.0,1,1.5,1.5,0,0,0,0",.... and these MBW values will be used instead of current MBW values,

for example,
when you select, MBW as "IN01, IN02, IN03" only, and select MBW alpha as "1,0, 1,1,1, 0,0,0,0,0,....0" as merge settings,
xyz MBW alpha B value with "1,0, 0.5,0.5, 0.5, 0", "1,0,0.2,0.2,0.2,0" will scan, IN01,IN02,IN03 == "0.5,0.5,0.5", "0.2,0.2,0.2"
(all other none zero MBW will be ignored except "IN01, IN02, IN03" in this case)


but it is not an easy task.

**Reply by miasik** [reply]
It's more difficult than I expected. I'd like to check only BASE block in three values. I set Mixer and ran. As I see in the log, it got the first line as an MBW string(but I'm not sure if it applied) and then it took each number as independent alpha, not as a multinumber MBW string.
![изображение](https://github.com/wkpark/sd-webui-model-mixer/assets/1073832/5df4a723-edb3-4676-877e-cd8369423012)
![изображение](https://github.com/wkpark/sd-webui-model-mixer/assets/1073832/dfd209f3-9d15-48a6-b25c-99edf57d7d92)
![изображение](https://github.com/wkpark/sd-webui-model-mixer/assets/1073832/9b0874de-0f59-4776-94a0-a839c9ec2059)
![изображение](https://github.com/wkpark/sd-webui-model-mixer/assets/1073832/5c6ca6ea-4140-4d32-99ab-31e816a2410f)

**Reply by wkpark** [reply]
MBW alpha B only accept 1 line. you should input "aaaa,aaa,aaa,aaa,","bbb,bbb,bbb,....","ccc,ccc,ccc,..."

**Reply by miasik** [reply]
It is already a line without breaks. It just looks like a multiline.
![изображение](https://github.com/wkpark/sd-webui-model-mixer/assets/1073832/5bb5365f-33a9-4fca-825c-7da4de0cc53c)

**Reply by wkpark** [reply]
ok. then, it could be a bug. I will check it soon

**Reply by wkpark** [reply]
XYZ plot with MBW alpha works as expected. you should select "Blocks" like follow
![image](https://github.com/wkpark/sd-webui-model-mixer/assets/232347/22a9fc8f-3eac-454f-94c1-9ee923cd49ca)

See also: https://github.com/wkpark/sd-webui-model-mixer/wiki/Merge-Block-Weights-UI

(MBW weight selection usage method is somewhat different from SuperMerger)

**Comment by wkpark** [comment]
and I recommend you to use "Auto Merger Helper" feature, It will automatically scan selected model or blocks.
XYZ scan is not an easy task, but experimental Auto Merger automatically scan and render a given prompt (or prompt from selected images with given path.)

**Reply by miasik** [reply]
I'd love to try it, but failed :-(
You should make an explanation about this feature because it's full of settings.

**Reply by wkpark** [reply]
If you don't mind, it would be nice to post a failure log as well.

**Comment by miasik** [comment]
> If you don't mind, it would be nice to post a failure log as well.

I failed on my own, not mixer :-)
Literally, I don't understand how it works.

**Comment by miasik** [comment]
> XYZ plot with MBW alpha works as expected. you should select "Blocks" like follow 

![изображение](https://github.com/wkpark/sd-webui-model-mixer/assets/1073832/ad399ce3-f149-4b14-8edf-1e01b5f1ac96)
![изображение](https://github.com/wkpark/sd-webui-model-mixer/assets/1073832/e8e58460-26f1-4b67-8d53-cd990d3a7b60)
Step 1 - MBW ok, wrong total steps, should be 3
![изображение](https://github.com/wkpark/sd-webui-model-mixer/assets/1073832/16bc2a71-72bf-4451-a0db-0f46ca05512a)
Step 2 - wrong
![изображение](https://github.com/wkpark/sd-webui-model-mixer/assets/1073832/8c943816-d8f6-4aca-a4c9-5b76a08dd3e2)

**Comment by wkpark** [comment, marked as answer]
do not input spaces between comma or quotes `","`
`"0.25,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0","0.5,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0","0.75,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0"`

not (_ denote a space)
`"0.25,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0",_"0.5,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0",_"0.75,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0"`

csv parser does not accept any space pre/post separator(default  separator is `,`)
~~~python
import csv
from io import StringIO

data_str1='"0.25,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0","0.5,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0","0.75,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0"'
data_str='"0.25,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0", "0.5,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0","0.75,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0"'
ll=csv.reader(StringIO(data_str))

print(list(ll))


~~~

**Comment by miasik** [comment]
> do not input spaces between comma or quotes `","` `"0.25,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0","0.5,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0","0.75,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0"`

It helped! Thanks so much!
